presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/IcebergPageSource.java

```diff
@@ -0,0 +1,254 @@
+                delegateIndexes[outputIndex] = -1;
+            }
+            else {
+                delegateIndexes[outputIndex] = delegateIndex;
+                delegateIndex++;
+            }
+            outputIndex++;
+        }
+    }
+
+    @Override
+    public long getCompletedBytes()
+    {
+        return delegate.getCompletedBytes();
+    }
+
+    @Override
+    public long getReadTimeNanos()
+    {
+        return delegate.getReadTimeNanos();
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return delegate.isFinished();
+    }
+
+    @Override
+    public Page getNextPage()
+    {
+        try {
+            Page dataPage = delegate.getNextPage();
+            if (dataPage == null) {
+                return null;
+            }
+            int batchSize = dataPage.getPositionCount();
+            Block[] blocks = new Block[prefilledBlocks.length];
+            for (int i = 0; i < prefilledBlocks.length; i++) {
+                if (prefilledBlocks[i] != null) {
+                    blocks[i] = prefilledBlocks[i].getRegion(0, batchSize);
+                }
+                else {
+                    blocks[i] = dataPage.getBlock(delegateIndexes[i]);
+                }
+            }
+            return new Page(batchSize, blocks);
+        }
+        catch (PrestoException e) {
+            closeWithSuppression(e);
+            throw e;
+        }
+        catch (RuntimeException e) {
+            closeWithSuppression(e);
+            throw new PrestoException(ICEBERG_BAD_DATA, e);
+        }
+    }
+
+    @Override
+    public void close()
+    {
+        try {
+            delegate.close();
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    @Override
+    public String toString()
+    {
+        return delegate.toString();
+    }
+
+    @Override
+    public long getSystemMemoryUsage()
+    {
+        return delegate.getSystemMemoryUsage();
+    }
+
+    protected void closeWithSuppression(Throwable throwable)
+    {
+        requireNonNull(throwable, "throwable is null");
+        try {
+            close();
+        }
+        catch (RuntimeException e) {
+            // Self-suppression not permitted
+            if (throwable != e) {
+                throwable.addSuppressed(e);
+            }
+        }
+    }
+
+    private static Object deserializePartitionValue(Type type, String valueString, String name, TimeZoneKey timeZoneKey)
+    {
+        try {
+            if (type.equals(BOOLEAN)) {
+                if (valueString.equalsIgnoreCase("true")) {
+                    return true;
+                }
+                if (valueString.equalsIgnoreCase("false")) {
+                    return false;
+                }
+                throw new IllegalArgumentException();
+            }
+            else if (type.equals(INTEGER)) {
```

Comment:
Nit: remove all these redundant "else" clauses (they are redundant because the "if" block always returns or throws)